stan/math/rev/mat/fun/adj_jac_apply.hpp

```diff
@@ -18,34 +35,224 @@ namespace math {
+    }
+    prepare_x_vis(args...);
+  }
+
+  template <int R, int C, typename... Pargs>
+  void prepare_x_vis(const Eigen::Matrix<double, R, C>& x,
+                     const Pargs&... args) {
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const std::vector<var>& x, const Pargs&... args) {
+    constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
+    for (size_t i = 0; i < x.size(); ++i) {
+      x_vis_[offsets_[t] + i] = x[i].vi_;
+    }
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const std::vector<double>& x, const Pargs&... args) {
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const std::vector<int>& x, const Pargs&... args) {
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const var& x, const Pargs&... args) {
+    constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
+    x_vis_[offsets_[t]] = x.vi_;
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const double& x, const Pargs&... args) {
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const int& x, const Pargs&... args) {
+    prepare_x_vis(args...);
+  }
+
+  void prepare_x_vis() {}
+
   /**
    * The adj_jac_vari constructor
    *  1. Initializes an instance of the user defined functor F
    *  2. Calls operator() on the F instance with the double values from the
-   * input x
+   * input args
    *  3. Saves copies of the varis pointed to by the input vars for subsequent
    * calls to chain
    *  4. Allocates varis for the output of the functor F
+   *  5. Populates the array is_var_ with whether or not the scalar type in each
+   * argument is a var
+   *
+   * Each of the arguments can be an Eigen::Matrix with var or double scalar
+   * types, a std::vector with var, double, or int scalar types, or a var, a
+   * double, or an int.
    *
-   * @param x Input as vars
+   * As an example of how is_var_ is initialized, take the adj_jac_vari
+   * constructed as: adj_jac_vari(a1, a2, a3, a4, a5)
+   *
+   * If a1 is an Eigen::Matrix<var, 5, 5>, a2 is a double, a3 is a var, a4 is
+   * a length 5 vector of vars, and a5 is a length 3 vector of ints,
+   * is_var_ is set to {true, false, true, true, false}
+   *
+   * @param args Input arguments
    */
-  explicit adj_jac_vari(const Eigen::Matrix<var, Eigen::Dynamic, 1>& x)
+  explicit adj_jac_vari(const Targs&... args)
       : vari(std::numeric_limits<double>::quiet_NaN()),  // The val_ in this
                                                          // vari is unused
-        N_(x.size()),
-        x_vi_(build_vari_array(x)) {
-    Eigen::Matrix<double, Eigen::Dynamic, 1> val_x = value_of(x);
-    Eigen::Matrix<double, Eigen::Dynamic, 1> val_y = f_(val_x);
+        is_var_({{is_var<typename scalar_type<Targs>::type>::value...}}),
```

Comment:
I still can't get over that these two lines will work.

Reply:
I didn't know what C++17 folds were and why I should care until I was playing with the include_summands stuff in the @dpsimpson thread:

```
value = (!propto || (!stan::is_constant<typename scalar_type<T_pack>::type>::value || ...))
```

Is fine hahaha.